Sara/reformat/scan engine kill (#1287)
semgrep/semgrep-docs: docs/kb/semgrep-code/scan-engine-kill.md

---
append_help_link: true
title: Troubleshoot monorepository scan failures
hide_title: true
description: Troubleshoot scan failures on monorepositories by studying logs, compartmentalizing scans, increasing RAM, and running jobs in parallel.
tags:
  - Semgrep Code
---

# Troubleshooting "You are seeing this because the engine was killed" on monorepositories

> Scans can fail to complete on large monorepositories. This article describes possible solutions, such as

Reviewer: ```suggestion
Scans can fail to complete on large monorepositories. This article describes possible solutions, such as:
```

- Scanning the components of the monorepository separately.
- Serializing the type of scan performed.

> - In CI jobs, increasing the RAM of the job runner.

Reviewer: ```suggestion
- Increasing the RAM of the job runner for CI jobs
```

          files. In this case, you can try to set the limit on the amount of memory
          semgrep can use on each file with `--max-memory <memory>`. We recommend
          lowering this to a limit 70% of the available memory. For CI runs with
          interfile analysis, the default max-memory is 5000MB. Without, the default
          is unlimited.

          The last thing you can try if none of these work is to raise the stack
          limit with `ulimit -s <limit>`.

          If you have tried all these steps and still are seeing this error, please
          contact us.

           Error: semgrep-core exited with unexpected output
```

## Determining the size of your monorepository 

By default, Semgrep places resource limitations on the size of file scanned and memory allocated.

However, Semgrep does not place limitations on the number of files scanned and scanning a large monorepository can involve thousands of lines of code. Thus, the first course of action is to determine exactly how many files are getting scanned:

1. View your CI logs. This step depends on your CI provider.
2. Search for the section "Scan Status".

> The following is an example of a large monorepository:

Reviewer: ```suggestion
The following is an example of a monorepository's components:
```

 ──────────────────────────────        ───────────────────
  <multilang>      60   199251          Community     551
  ts              166    26672          Pro rules     498
  python          314     8089          Custom         19
  scala            13     5415
  json              1     4149       
  yaml              7     1952
  js              160     1084      
  terraform        13      470      
  bash              4      408           
  go               95      228
  ruby             22      228
  php              25       99
  swift            41       89
  html              1       72         
  dockerfile        2       31           
  c                 9       16
  rust             49        8
  java            156        2
  kotlin           47        1
```

Now you have a good idea of the size of your monorepository. After establishing the size and breakdown of programming languages and files, you can decide what adjustments to take for a scan to succeed.

## Scanning components separately 

Based on the composition provided by the logs, you may be able to determine if your repository is modular. If so, you can try scanning the components separately.

> Interfile analysis still applies in the smaller scope assuming the modules are disparate. 

Reviewer: maybe a link to a definition of interfile analysis here?

Author: Ye! Rather bare of links, this article is.

### Scanning schedule

You can perform modular scans on a daily basis and then perform one big monolithic scan every two weeks or so to keep close tabs and cap any false negative interfile or inter-component outliers that can get introduced with code change as a way of truing up your dependency graph for the monolithic build. This helps deflect costs as you are scanning the majority on smaller runners/k8s.  

However, you still need to scan the whole monolithic build, whether you scan it biweekly or daily, and the following sections apply to tackling that monolith whether you adopt the modular paradigm of scanning or not. 

## Serializing types of scans

With Semgrep 1.42.0 and later, Secret rulesets are now included in the default scan configuration.  Instead of scanning Code, Supply Chain and Secret rulesets concurrently, you can circumnavigate resource limits by running different types of scans in a serial fashion.  Rather than issuing one monolithic command, for example:

```console
 $ semgrep ci

> ```

Reviewer: small nit: do you think we should omit the $ from the codeblock so that users can copy and paste without having to modify what they've copied before running?

Author: Good idea. I'll make the change.

 
You can instead run:
 
 ```
$ semgrep ci --code --pro
$ semgrep ci --supply-chain --oss-only
$ semgrep ci --secrets
 
 ```

As a result, less memory is used in total at any point in time.

## Increasing RAM 

Lastly, you can also tackle a large scan by increasing the RAM.

### Establish RAM baseline and avoid swap memory

> First, establish how much memory is required to scan. Determining the total amount of memory required is key, not just to avoid killed scans but running scans that use swap memory. Avoid swap memory when scanning with any static analysis tool. Semgrep and other SAST tools are disk I/O intensive and swapping in and out with a swap file reduces performance severely.

Reviewer: ```suggestion
First, establish how much memory is required to scan. Determining the total amount of memory required is key, not just to avoid killed scans but to run scans that use swap memory. Avoid swap memory when scanning with any static analysis tool. Semgrep and other SAST tools are disk I/O intensive and swapping in and out with a swap file reduces performance severely.
```
